src/include/stat.h

```diff
@@ -6,35 +6,169 @@
+ * a heuristic for the number of CPUs. Unfortunately, However, inside WT we do
+ * not know when the application creates its threads.
+ *
+ * Our solution is to simply use a fixed number of slots that roughly
+ * approximates the largest number of cores we expect to see on the machine
+ * where WT is used. All we want is to avoid all threads writing the same
+ * variable at the same time, and that solution does the trick.
+ *
+ */
+#define	WT_COUNTER_SLOTS 24
+
+/* This is a cache-line-padded counter value. 
+ * It is here to effectively support array counters.
+ * Padding is needed, otherwise cache coherency messages
+ * will be triggered because of false sharing.
+ *
+ * The actual counter value v must be signed, because it is possible
+ * that one thread incremented the counter in its own slot, and then
+ * another thread decremented the same counter in another slot,
+ * which was initially zero. We need to allow the value is the second
+ * thread's slot to be negative. When the values are aggregated, we get
+ * the correct total value.
+ */
+typedef struct{
+	int64_t v;
+	char padding[WT_CACHE_LINE_ALIGNMENT-sizeof(uint64_t)];
+} padded_counter_t;
+
 struct __wt_stats {
 	const char	*desc;				/* text description */
-	uint64_t	 v;				/* 64-bit value */
+//  	uint64_t	 v;				/* 64-bit value */
+	padded_counter_t array_v[WT_COUNTER_SLOTS];     /* Array counter */
 };
 
 /*
+ * This macro determines the slot id for the array of counters.
+ *
+ * Ideally, we want a slot per CPU, and we want each thread to index the slot
+ * corresponding to the CPU it runs on. Unfortunately, getting the id of the
+ * current CPU is difficult. Some operating systems provide a system call for
+ * that, but not all. Further, invoking a system call every time we need to
+ * increment a stats counter is expensive. Another option would be to use the
+ * rdtscp instruction, but it's x86 specific, has a ~50 cycle overhead and is
+ * not supported on some older processors.
+ *
+ * Our second-best option is to use the thread ID. Unfortunately, there is no
+ * portable way to obtain a thread ID that is a small-enough number that could
+ * be used as an array index.
+ *
+ * Our solution is to use the session ID, because there is a session per thread
+ * and the id itself is a small monotonically increasing number, so this results
+ * in threads writing into different array slots, which is ultimately what we
+ * want.
+ */
+#define	WT_STATS_SLOT_ID	(session->id) % WT_COUNTER_SLOTS
+
+/* Set all the values in the array counter slots to zero. We are
+ * doing more work than needed by using memset, because we are setting
+ * both the values and the padding. However, resetting the counters is
+ * not a common case operation, so we use memset for compactness.
+ */
+#define	WT_STAT_ALL_RESET(stats, fld)  do {				\
+		memset((stats)->fld.array_v, 0,                         \
+		       sizeof(padded_counter_t) * WT_COUNTER_SLOTS);	\
+} while (0)
```

Comment:
Keith, 

My intuition fails me here. I’ll run tests on a couple of machines this week-end to get more insight. 

> On Aug 12, 2015, at 9:44 AM, Keith Bostic <notifications@github.com> wrote:
> 
> In src/include/stat.h <https://github.com/wiredtiger/wiredtiger/pull/2102#discussion_r36882984>:
> 
> > + * Our solution is to use the session ID, because there is a session per thread
> > + * and the id itself is a small monotonically increasing number, so this results
> > + * in threads writing into different array slots, which is ultimately what we
> > + * want.
> > + */
> > +#define	WT_STATS_SLOT_ID	(session->id) % WT_COUNTER_SLOTS
> > +
> > +/* Set all the values in the array counter slots to zero. We are
> > + * doing more work than needed by using memset, because we are setting
> > + * both the values and the padding. However, resetting the counters is
> > + * not a common case operation, so we use memset for compactness.
> > + */
> > +#define	WT_STAT_ALL_RESET(stats, fld)  do {				\
> > +		memset((stats)->fld.array_v, 0,                         \
> > +		       sizeof(padded_counter_t) * WT_COUNTER_SLOTS);	\
> > +} while (0)
> @fedorova <https://github.com/fedorova>, applications can configure statistics=clear to wiredtiger_open, which resets statistics after they are gathered, where appropriate (for example, a cache size statistic is not cleared, while the count of cursor insert operations will be cleared). So, those applications are clearing the stats every time they're gathered.
> 
> It's probably simplest to build a standalone C program that just compares memset to clearing each value, and run it for a million loops? I'm also suspicious that different compilers and different library implementations of memset (and even different underlying instruction sets on different x86 and other architectures, we're not limited to the x86), will come into play.
> 
> Since figuring out the "right" answer sounds like a ton of work, and I doubt it's worth that much effort, I'd be happy to go with your best guess on this one for now -- if you think memset is likely faster, we'll go with it, if not, we can write the inline function to clear the values.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/wiredtiger/wiredtiger/pull/2102/files#r36882984>.
>